x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,515 @@
+        
+        @Override
+        protected void setupFieldType(BuilderContext context) {
+            super.setupFieldType(context);
+            fieldType().setNumChars(numChars);
+            fieldType().setHasDocValues(true);
+            fieldType().setIndexOptions(IndexOptions.DOCS);                
+        }
+
+        @Override
+        public WildcardFieldType fieldType() {
+            return (WildcardFieldType) super.fieldType();
+        }
+
+        @Override
+        public WildcardFieldMapper build(BuilderContext context) {
+            setupFieldType(context);            
+            return new WildcardFieldMapper(
+                    name, fieldType, defaultFieldType, ignoreAbove, numChars,
+                    context.indexSettings(), multiFieldsBuilder.build(this, context), copyTo);
+        }
+    }
+
+    public static class TypeParser implements Mapper.TypeParser {
+        @Override
+        public Mapper.Builder<?, ?> parse(String name, Map<String, Object> node, ParserContext parserContext)
+                throws MapperParsingException {
+            WildcardFieldMapper.Builder builder = new WildcardFieldMapper.Builder(name);
+            parseField(builder, name, node, parserContext);
+            
+            for (Iterator<Map.Entry<String, Object>> iterator = node.entrySet().iterator(); iterator.hasNext();) {
+                Map.Entry<String, Object> entry = iterator.next();
+                String propName = entry.getKey();
+                Object propNode = entry.getValue();
+                if (propName.equals("num_chars")) {
+                    if (propNode == null) {
+                        throw new MapperParsingException("Property [numChars] cannot be null.");
+                    }
+                    builder.numChars(XContentMapValues.nodeIntegerValue(propNode));
+                    iterator.remove();
+                } else if (propName.equals("ignore_above")) {
+                    builder.ignoreAbove(XContentMapValues.nodeIntegerValue(propNode, -1));
+                    iterator.remove();
+                }
+            }            
+            
+            return builder;
+        }
+    }
+    
+     public static final char TOKEN_START_OR_END_CHAR = 0;
+    // A visible character to aid debug
+//     public static final char TOKEN_START_OR_END_CHAR = '$';
+    
+    
+     public static final class WildcardFieldType extends MappedFieldType {
+        private int numChars;
+
+        public WildcardFieldType() {            
+            setIndexAnalyzer(Lucene.KEYWORD_ANALYZER);
+            setSearchAnalyzer(Lucene.KEYWORD_ANALYZER);            
+        }
+
+        protected WildcardFieldType(WildcardFieldType ref) {
+            super(ref);
+        }
+
+        public WildcardFieldType clone() {
+            return new WildcardFieldType(this);
+        }
+        
+        
+        // Holds parsed information about the wildcard pattern
+        static class PatternStructure {
```

Comment:
I haven't looked deeply in this logic, but do you think it would be much harder to extract ngrams from an Automaton? I'm asking because this would give this field support for wildcard, fuzzy and regexp queries.

Reply:
Avoiding our own parsing would be ideal but I'll need to dig deeper into the Automaton structures to understand if we can get out what we need. One challenge is we don't have fix-sized n-grams (the end of a string tapers down from 3 to 2 to 1 ngrams). If we have a pattern like `*.c` we know we can query for just `c` with the end-of-line char added because we know that's in the index but `*.c*` would require a `c*` wildcard query on the ngrams index. I'm not sure how much of that information I can glean from the automaton.

Reply:
Theoretically, the automaton has all the information, but I haven't looked into how hard it was to convert it to a set of ngrams.